### PR TITLE
[nccl] Remove NCCL auto-download due to licensing issues

### LINKED ISSUE #950: Installation Failed: flashlight-{cuda/cpu}[asr]
I am installing on the brand-new google compute engine with Ubuntu 2004 and and Tesla V100.  All my previous installation steps are successfully (MKL, Cuda - 11.2, Cudnn 8.1, nccl-2.8.4) and vcpkg. I actually have successfully finished the step:

$ ./vcpkg install flashlight-cuda

But when I ran the next step

$ ./vcpkg install flashlight-cuda[asr]

The 'vcpkg' always prompts me that I need to pass the '--recursive' flag to have the [core,lib,asr] all built, and when I actually did that, the build failed with an error as following:

[186/193] : && /usr/bin/c++ -fPIC -g -rdynamic CMakeFiles/fl_asr_tutorial_inference_ctc.dir/flashlight/app/asr/tutorial/InferenceCTC.cpp.o -o bin/asr/fl_asr_tutorial_inference_ctc -L/usr/local/cuda-11.2/targets/x86_64-linux/lib/stubs   -L/usr/local/cuda-11.2/targets/x86_64-linux/lib -Wl,-rpath,/home/xiaopanzhang/vcpkg/packages/arrayfire_x64-linux/lib:/home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib:/home/xiaopanzhang/vcpkg/installed/x64-linux/lib  libflashlight-app-asr.a  -ldl  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libglog.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libgflags.a  libflashlight.a  libfl-libraries.a  /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_intel_lp64.so  /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_gnu_thread.so  /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_core.so  -fopenmp  /usr/lib/x86_64-linux-gnu/libpthread.so  -lm  /usr/lib/x86_64-linux-gnu/libdl.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libfftw3.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libkenlm.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libkenlm_util.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/liblzmad.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libbz2d.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libz.a  /home/xiaopanzhang/vcpkg/packages/arrayfire_x64-linux/lib/libafcuda.so.3.7.3  -pthread  /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libcudnn.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libmpi.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libnccl.so  /usr/local/cuda-11.2/lib64/libcudart_static.a  -ldl  -lpthread  /usr/lib/x86_64-linux-gnu/librt.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libsndfile.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libvorbisenc.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libvorbis.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libFLAC.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libogg.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libopus.a  -lm  -lcudadevrt  -lcudart_static  -lrt  -lpthread  -ldl && :
FAILED: bin/asr/fl_asr_tutorial_inference_ctc
: && /usr/bin/c++ -fPIC -g -rdynamic CMakeFiles/fl_asr_tutorial_inference_ctc.dir/flashlight/app/asr/tutorial/InferenceCTC.cpp.o -o bin/asr/fl_asr_tutorial_inference_ctc -L/usr/local/cuda-11.2/targets/x86_64-linux/lib/stubs   -L/usr/local/cuda-11.2/targets/x86_64-linux/lib -Wl,-rpath,/home/xiaopanzhang/vcpkg/packages/arrayfire_x64-linux/lib:/home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib:/home/xiaopanzhang/vcpkg/installed/x64-linux/lib  libflashlight-app-asr.a  -ldl  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libglog.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libgflags.a  libflashlight.a  libfl-libraries.a  /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_intel_lp64.so  /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_gnu_thread.so  /opt/intel/oneapi/mkl/latest/lib/intel64/libmkl_core.so  -fopenmp  /usr/lib/x86_64-linux-gnu/libpthread.so  -lm  /usr/lib/x86_64-linux-gnu/libdl.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libfftw3.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libkenlm.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libkenlm_util.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/liblzmad.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libbz2d.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libz.a  /home/xiaopanzhang/vcpkg/packages/arrayfire_x64-linux/lib/libafcuda.so.3.7.3  -pthread  /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libcudnn.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libmpi.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libnccl.so  /usr/local/cuda-11.2/lib64/libcudart_static.a  -ldl  -lpthread  /usr/lib/x86_64-linux-gnu/librt.so  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libsndfile.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libvorbisenc.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libvorbis.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libFLAC.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libogg.a  /home/xiaopanzhang/vcpkg/installed/x64-linux/debug/lib/libopus.a  -lm  -lcudadevrt  -lcudart_static  -lrt  -lpthread  -ldl && :
/usr/bin/ld: warning: libcuda.so.1, needed by /home/xiaopanzhang/vcpkg/packages/arrayfire_x64-linux/lib/libafcuda.so.3.7.3, not found (try using -rpath or -rpath-link)
/usr/bin/ld: warning: libcudart.so.10.1, needed by /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libnccl.so, not found (try using -rpath or -rpath-link)
/usr/bin/ld: /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libnccl.so: undefined reference to `cudaGetDeviceCount@libcudart.so.10.1'
/usr/bin/ld: /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libnccl.so: undefined reference to `__cudaRegisterFunction@libcudart.so.10.1'
/usr/bin/ld: /home/xiaopanzhang/vcpkg/installed/x64-linux/lib/libnccl.so: undefined reference to `cudaIpcCloseMemHandle@libcudart.so.10.1'

As far as I can see, the link failed because it looks like the linker is looking for dynamic libraries which is different from the version I installed, but I do have LD_LIBRARY_PATH and PATH point to the cuda installation location (/usr/local/cuda-11.2/... here), I also check the library list of 'vcpkg' and it says:

xiaopanzhang@wav2letter:~/vcpkg$ ./vcpkg list | grep cuda
arrayfire[cuda]:x64-linux                                           ArrayFire CUDA backend
cuda:x64-linux                                     10.1#5           A parallel computing platform and programming model

xiaopanzhang@wav2letter:~/vcpkg$ ./vcpkg list | grep nccl
nccl:x64-linux                                     2.4.6            Optimized primitives for collective multi-GPU co..

It looks like the vcpkg also has its own old version of cuda, cudnn and nccl, could anybody tell me what might go wrong here?

### PR DESCRIPTION
**Pull request by jacobkahn**
- Pursuant to https://github.com/microsoft/vcpkg/pull/17331, remove NCCL auto-download due to licensing issues.
- Also improves finding NCCL - uses common environment variables

- #### What does your PR fix?  

Follow up to https://github.com/microsoft/vcpkg/pull/16031, mimics https://github.com/microsoft/vcpkg/pull/16413, and makes additional changes as outlined in https://github.com/microsoft/vcpkg/pull/17346

Fixes https://github.com/flashlight/wav2letter/issues/950

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  

Same triplets as before. This doesn't affect the CI baseline since CUDA 10.1 is still installed on CI Machines and NCCL 2.4.6 (compatible with 10.1) will still be downloaded from conda.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  

Yes

### CODE REVIEW

**Comment by cenit** [comment]
please consider this:
https://github.com/microsoft/vcpkg/pull/17331#issuecomment-824302062

**Comment by jacobkahn** [comment]
Aha - thanks for flagging @cenit -- I'll go ahead and remove it here too for this PR. cc @BillyONeal — if you want to tack NCCL onto [your request](https://twitter.com/MalwareMinigun/status/1384335333249097732), that would be helpful!

Too bad that we have to remove it, but makes sense.

File: ports/nccl/portfile.cmake [hunk 1]
```diff
@@ -1,55 +1,21 @@
...
+find_package(NCCL ${MINIMUM_NCCL_VERSION})
 set(CMAKE_MODULE_PATH ${NCCL_PREV_MODULE_PATH})
 
 # Download or return
 if(NCCL_FOUND)
-  message(STATUS "Using NCCL located on system.")
+  message(STATUS "Using NCCL ${_NCCL_VERSION} located on system.")
   set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
+elseif(VCPKG_TARGET_IS_WINDOWS)
+  message(FATAL_ERROR "Please download NCCL from official sources (such as https://developer.nvidia.com/nccl ) and extract the zip into your CUDA_TOOLKIT_ROOT (${CUDA_TOOLKIT_ROOT})")
```
Comment (BillyONeal) [inline ports/nccl/portfile.cmake:16]: Can we print the actual tar.exe command the user needs in order to get the right results here?

Comment (jacobkahn) [reply ports/nccl/portfile.cmake:16]: Yep -- I can try to find out what it should be...

Comment (jacobkahn) [reply ports/nccl/portfile.cmake:16]: Also @BillyONeal — any change we could get a NCCL version compatible with CUDA 11.3 on the CI machines as well? We have > 1 ports that will depend on it...

Comment (BillyONeal) [reply ports/nccl/portfile.cmake:16]: Ah nvm this one is Linux only so no tar command required.

**Comment by BillyONeal** [comment]
Will have this pending shortly: 
image

**Comment by PhoebeHui** [comment]
Depends on PR https://github.com/microsoft/vcpkg/pull/17331.

Failures on linux:
```
-- Could NOT find NCCL: Found unsuitable version "", but required is at least "2.4.6.1" (found NCCL_INCLUDE_DIRS-NOTFOUND)
CMake Error at ports/nccl/portfile.cmake:16 (message):
  Please install NCCL using your system package manager (the same way you
  installed CUDA).  For example: apt install libnccl2 libnccl-dev.
Call Stack (most recent call first):
  scripts/ports.cmake:142 (include)
```

**Comment by PhoebeHui** [comment]
@jacobkahn, thanks for your contribution!

Could you update the baseline version by executing ' ./vcpkg x-add-version --overwrite-version nccl'?

```
Error: While reading versions for port nccl from file: C:\a\1\s\versions\n-\nccl.json
       File declares version `2.4.6#1` with SHA: 78f69d23214c0c0849fbe1a5d0fe62f934dc72e1
       But local port with the same verion has a different SHA: 65d03efc8a527afe223125d2cac2e685517297e3
       Please update the port's version fields and then run:

           vcpkg x-add-version nccl

       to add a new version.
```

**Comment by jacobkahn** [comment]
@PhoebeHui done.

**Comment by BillyONeal** [comment]
> I think CI needs a rerun now that NCCL is installed.

You can always trigger that by merging with `origin/master`.

**Review by BillyONeal — DISMISSED** [review]
Looks OK to me assuming CI passes.

**Comment by BillyONeal** [comment]
Looks like CI failures are a regression from https://github.com/microsoft/vcpkg/pull/13100 + updating the compilers on the build machines to 16.9.4

**Comment by PhoebeHui** [comment]
Yes, For gdk-pixbuf's failure, it's not relate to this PR, we would take the baseline error.

Failures:
```
E:\vcpkg\clean\vcpkg\buildtrees\gdk-pixbuf\src\k-pixbuf-2-c86079bed6\gdk-pixbuf\pixops\../fallback-c89.c(38): error C2169: 'lrint': intrinsic function, cannot be defined
```

For nccl:x64-linux, it still failed with following failure:
```
-- Could NOT find NCCL: Found unsuitable version "", but required is at least "2.4.6.1" (found NCCL_INCLUDE_DIRS-NOTFOUND)
CMake Error at ports/nccl/portfile.cmake:16 (message):
  Please install NCCL using your system package manager (the same way you
  installed CUDA).  For example: apt install libnccl2 libnccl-dev.
Call Stack (most recent call first):
  scripts/ports.cmake:142 (include)
```

**Comment by Neumann-A** [comment]
gdk-pixbuf check for lrint in meson.build needs to be updated to use the correct parameters. I often see function checks in the form of `somefunction(void)` and VS is eagerly reporting errors with release build flags while with debug build flags the check passes fine.

**Comment by jacobkahn** [comment]
Looks like the dependent PR has merged -- are we good here @BillyONeal @PhoebeHui ?

**Comment by PhoebeHui** [comment]
@jacobkahn, I have rerun the CI testing, let's wait for the results.

**Comment by BillyONeal** [comment]
> Looks like the dependent PR has merged -- are we good here @BillyONeal @PhoebeHui ?

This one looks like the port actually not being happy:

```
-- Could NOT find NCCL: Found unsuitable version "", but required is at least "2.4.6.1" (found NCCL_INCLUDE_DIRS-NOTFOUND)
CMake Error at ports/nccl/portfile.cmake:16 (message):
  Please install NCCL using your system package manager (the same way you
  installed CUDA).  For example: apt install libnccl2 libnccl-dev.
Call Stack (most recent call first):
  scripts/ports.cmake:142 (include)
```

We installed nccl into our base VMs like this: https://github.com/microsoft/vcpkg/blob/1efdaf8f4da22f2825567cd88d7c10b60a585a1f/scripts/azure-pipelines/linux/provision-image.sh#L66

**Comment by jacobkahn** [comment]
@BillyONeal reprod and fixed. It seems that like cuDNN, NCCL throws bits directly in `/usr/include` and `/usr/lib`, and there's no pkgconfig, so added those hints as we have for the cuDNN package.

**Review by PhoebeHui — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
Thanks @jacobkahn and @BillyONeal !

**Comment by longnguyen2004** [comment]
Can we also build nccl from source? The source is at https://github.com/NVIDIA/nccl/tree/master/src

**Comment by BillyONeal** [comment]
@longnguyen2004 I would really like to get some form of nvidia comment before going there; the licensing in that repo is not consistent.

**Comment by longnguyen2004** [comment]
Seems to me that it's mostly BSD 3-clause, judging from this page https://docs.nvidia.com/deeplearning/nccl/bsd/index.html

**Pull request merged**